packages/generate/src/codecToType.ts

```diff
@@ -0,0 +1,152 @@
+import type {ParseResult} from "edgedb/dist/baseConn";
+import {ArrayCodec} from "edgedb/dist/codecs/array";
+import {AT_LEAST_ONE, AT_MOST_ONE, MANY, ONE} from "edgedb/dist/codecs/consts";
+import {EnumCodec} from "edgedb/dist/codecs/enum";
+import {ICodec, ScalarCodec} from "edgedb/dist/codecs/ifaces";
+import {NamedTupleCodec} from "edgedb/dist/codecs/namedtuple";
+import {ObjectCodec} from "edgedb/dist/codecs/object";
+import {RangeCodec} from "edgedb/dist/codecs/range";
+import {NullCodec} from "edgedb/dist/codecs/codecs";
+import {SetCodec} from "edgedb/dist/codecs/set";
+import {TupleCodec} from "edgedb/dist/codecs/tuple";
+import {Cardinality, OutputFormat} from "edgedb/dist/ifaces";
+import {Options, Session} from "edgedb/dist/options";
+import type {Client} from "edgedb";
+import type {ClientPool} from "edgedb/dist/client";
+import {prettyPrintError} from "./prettyPrint";
+
+export type QueryType = {
+  args: string;
+  out: string;
+  cardinality: Cardinality;
+  query: string;
+  imports: Set<string>;
+};
+export async function generateQueryType(
+  client: Client,
+  query: string
+): Promise<QueryType> {
+  let parseResult: ParseResult;
+  const pool: ClientPool = (client as any).pool;
+
+  const holder = await pool.acquireHolder(Options.defaults());
+  try {
+    const cxn = await holder._getConnection();
+    parseResult = await cxn._parse(
+      query,
+      OutputFormat.BINARY,
+      Cardinality.MANY,
+      Session.defaults(),
+      false
+    );
+  } catch (err) {
+    throw new Error(prettyPrintError(err, query));
+  } finally {
+    await holder.release();
+  }
+
+  const cardinality = parseResult[0];
+  const inCodec = parseResult[1];
+  const outCodec = parseResult[2];
+  const imports = new Set<string>();
+  const args = walkCodec(inCodec, {
+    indent: "",
+    optionalNulls: true,
+    imports
+  });
+
+  const out = generateSetType(
+    walkCodec(outCodec, {
+      indent: "",
+      optionalNulls: false,
+      imports
+    }),
+    cardinality
+  );
+
+  return {
+    out,
+    args,
+    cardinality,
+    query,
+    imports
+  };
+}
+
+function generateSetType(type: string, cardinality: Cardinality): string {
+  switch (cardinality) {
+    case Cardinality.MANY:
+      return `${type}[]`;
+    case Cardinality.ONE:
+      return type;
+    case Cardinality.AT_MOST_ONE:
+      return `${type} | null`;
+    case Cardinality.AT_LEAST_ONE:
+      return `AtLeastOne<${type}>`;
```

Comment:
We need to import this `AtLeastOne` type when it's used.

Reply:
I inlined this instead